### PR TITLE
Find cases by samples

### PR DESCRIPTION
**Pull request by damien-lloyd**
# Motivation and Context
We want to find cases that relate to samples uploaded via a sample file. To do this, we search by the sample's post code and this will return cases. 

# What has changed
Endpoint for returning cases based on sample unit id

# How to test?
Run all integration tests

# Links
https://trello.com/c/uqLRhZiZ/189-sus002-find-case-by-postcode-13#

### CODE REVIEW

**Comment by damien-lloyd** [comment]
[SUS002: Find case by Postcode (13)](https://trello.com/c/uqLRhZiZ/189-sus002-find-case-by-postcode-13)

File: src/main/java/uk/gov/ons/ctp/response/casesvc/domain/repository/CaseRepository.java [hunk 1]
```diff
@@ -85,4 +87,15 @@
...
+  /**
+   * Find case by its sample unit id
+   *
+   * @param sampleUnitId
+   * @return
+   */
+  @Query(
+      value = "select caze.* from casesvc.case caze where sampleunit_id = :sampleUnitId",
+      nativeQuery = true)
+  Case findBySampleUnitId(@Param("sampleUnitId") UUID sampleUnitId);
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/domain/repository/CaseRepository.java]: This probably doesn't need to be a native query and we could possibly use the in clause https://docs.spring.io/spring-data/jpa/docs/1.5.0.RC1/reference/html/repository-query-keywords.html

File: src/main/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseEndpoint.java [hunk 2]
```diff
@@ -103,6 +104,29 @@ public CaseEndpoint(
...
+      @RequestParam(value = "sampleUnitId") final UUID[] sampleUnitIds,
+      @RequestParam(value = "caseevents", required = false) final boolean caseevents,
+      @RequestParam(value = "iac", required = false) final boolean iac)
+      throws CTPException {
+
+    List<UUID> ids = Arrays.asList(sampleUnitIds);
+
+    List<CaseDetailsDTO> cases =
+        ids.stream()
+            .filter(e -> caseService.findCaseBySampleUnitId(e) != null)
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseEndpoint.java]: We could possible use the in clause to do this as one query https://docs.spring.io/spring-data/jpa/docs/1.5.0.RC1/reference/html/repository-query-keywords.html

Comment (benjefferies) [reply src/main/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseEndpoint.java]: We should also map first so we don't have to do the query twice per ID

File: src/main/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseEndpoint.java [hunk 3]
```diff
@@ -103,6 +104,29 @@ public CaseEndpoint(
     return ResponseEntity.ok(buildDetailedCaseDTO(caseObj, caseevents, iac));
   }
 
+  @RequestMapping(value = "/sampleunitids", method = RequestMethod.GET)
+  public ResponseEntity<CaseDetailsDTO[]> findCasesBySampleUnitIds(
+      @RequestParam(value = "sampleUnitId") final UUID[] sampleUnitIds,
```
Comment (atrozzo) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseEndpoint.java]: why this one is not directly a List ?

File: src/main/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseEndpoint.java [hunk 4]
```diff
@@ -103,6 +104,29 @@ public CaseEndpoint(
...
+      @RequestParam(value = "caseevents", required = false) final boolean caseevents,
+      @RequestParam(value = "iac", required = false) final boolean iac)
+      throws CTPException {
+
+    List<UUID> ids = Arrays.asList(sampleUnitIds);
+
+    List<CaseDetailsDTO> cases =
+        ids.stream()
```
Comment (atrozzo) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseEndpoint.java]: You could have done directly Arrays.stream(array)

File: src/main/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseEndpoint.java (hunk 3 above)
Comment (benjefferies) [reply src/main/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseEndpoint.java]: This makes sense, then we don't need to do anything else to stream it

File: src/main/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseEndpoint.java [hunk 5]
```diff
@@ -103,6 +103,27 @@ public CaseEndpoint(
...
+  @RequestMapping(value = "/sampleunitids", method = RequestMethod.GET)
+  public ResponseEntity<List<CaseDetailsDTO>> findCasesBySampleUnitIds(
+      @RequestParam(value = "sampleUnitId") final List<UUID> sampleUnitIds,
+      @RequestParam(value = "caseevents", required = false) final boolean caseevents,
+      @RequestParam(value = "iac", required = false) final boolean iac) {
+
+    List<CaseDetailsDTO> cases =
+        sampleUnitIds
+            .stream()
+            .filter(e -> caseService.findCaseBySampleUnitId(e) != null)
```
Comment (benjefferies) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/endpoint/CaseEndpoint.java]: If we do a map here followed by the filter we only need to do the DB lookup once

**Pull request merged**